core/common/assets/js/api/core/data.js

```diff
@@ -0,0 +1,430 @@
+			endpoint = endpoint.substring( 0, endpoint.indexOf( '/:' ) );
+		}
+
+		return endpoint;
+	}
+
+	/**
+	 * Function endpointToCommand().
+	 *
+	 * Convect endpoint to command.
+	 *
+	 * @param {string} endpoint
+	 * @param {object} args
+	 *
+	 * @return {string} command
+	 */
+	endpointToCommand( endpoint, args = {} ) {
+		let command = endpoint,
+			commandFound = !! $e.data.commands[ endpoint ];
+
+		const endpointParts = endpoint.split( '/' );
+
+		// Assuming the command maybe index.
+		if ( ! commandFound && $e.data.commands[ endpoint + '/index' ] ) {
+			command = endpoint + '/index';
+			commandFound = true;
+		}
+
+		// Maybe the it has 'id' as last endpointPart.
+		if ( ! commandFound ) {
+			// Remove last parameter.
+			const lastParameter = endpointParts[ endpointParts.length - 1 ],
+				assumedCommand = command.replace( '/' + lastParameter, '' );
+
+			// If assumed command ( without last parameter ) exist.
+			if ( $e.data.commands[ assumedCommand ] ) {
+				command = assumedCommand;
+				commandFound = true;
+
+				if ( ! args.query ) {
+					args.query = {};
+				}
+
+				// Warp with 'id'.
+				args.query.id = lastParameter;
+			}
+		}
+
+		return command;
+	}
+
+	/**
+	 * Function prepareHeaders().
+	 *
+	 * @param {DataTypes} type
+	 * @param {RequestData} requestData
+	 *
+	 * @return {{}} params
+	 */
+	prepareHeaders( type, requestData ) {
+		/* global wpApiSettings */
+		const nonce = wpApiSettings.nonce,
+			params = {
+				credentials: 'include', // cookies is required for wp reset.
+			},
+			headers = { 'X-WP-Nonce': nonce };
+
+		/**
+		 * Translate:
+		 * 'create, delete, get, update' to HTTP Methods:
+		 * 'GET, POST, PUT, PATCH, DELETE'
+		 */
+		const allowedMethods = this.getAllowedMethods( type ),
+			method = this.getHTTPMethod( type );
+
+		if ( 'GET' === method ) {
+			Object.assign( params, { headers } );
+		} else if ( allowedMethods ) {
+			Object.assign( headers, { 'Content-Type': 'application/json' } );
+			Object.assign( params, {
+				method,
+				headers: headers,
+				body: JSON.stringify( requestData ),
+			} );
+		} else {
+			throw Error( `Invalid type: '${ type }'` );
+		}
+
+		return params;
+	}
+
+	/**
+	 * Function fetch().
+	 *
+	 * @param {DataTypes} type
+	 * @param {RequestData} requestData
+	 *
+	 * @return {{}} params
+	 */
+	fetch( type, requestData ) {
+		const params = this.prepareHeaders( type, requestData ),
+			useCache = 'get' === type && ! requestData.args.options?.refresh;
+
+		if ( useCache ) {
+			const cachePromise = this.cache.receive( requestData );
+
+			if ( cachePromise ) {
+				return cachePromise;
+			}
+		}
+
+		return new Promise( async ( resolve, reject ) => {
+			try {
+				const request = window.fetch( this.baseEndpointAddress + requestData.endpoint, params );
+
+				let response = await request.then();
+
+				await request.catch( reject );
+
+				if ( ! response.ok ) {
+					throw response;
+				}
+
+				// Ensure JSON.
+				response = await response.json();
```

Comment:
cannot do before throw, since i need to validate:
```
				if ( ! response.ok ) {
					throw response;
				}

```

Reply:
This can work?
```
let response = await request
.then( ( response ) => {
	if ( ! response.ok ) {
		throw response;
	}

	return response.json();
} )
.catch( reject );
```

Reply:
Handled.